### PR TITLE
Refactor normalize API

### PR DESCRIPTION
**Pull request by steff456**
## Description

> What is the purpose of this pull request?

This pull request:

-   Refactors `normalize` package to support `assign` when the result needs to be overwrite a given array

## Questions

> Any questions for reviewers of this pull request?

No.

## Other

> Any other information relevant to this pull request? This may include screenshots, references, and/or implementation notes.

For some reason my node is flushing subnormals to 0, so I couldn't test the function in some cases. I will try to run them on another machine, but someone else double checking them will be nice :)!

## Checklist

> Please ensure the following tasks are completed before submitting this pull request.

-   [x] Read, understood, and followed the [contributing guidelines][contributing].

* * *

@stdlib-js/reviewers

[contributing]: https://github.com/stdlib-js/stdlib/blob/develop/CONTRIBUTING.md

### CODE REVIEW

File: lib/node_modules/@stdlib/number/float64/base/normalize/docs/types/index.d.ts [hunk 1]
```diff
@@ -23,77 +23,85 @@
...
+	* Returns a normal number `y` and exponent `exp` satisfying \\(x = y \cdot 2^\mathrm{exp}\\).
+	*
+	* The first element of the returned array corresponds to `y` and the second to `exp`.
+	*
+	* @param x - input value
+	* @returns output array
+	*
+	* @example
+	* var pow = require( `@stdlib/math/base/special/pow` );
+	* var out = normalize( 3.14e-319 );
```
Comment (kgryte) [inline lib/node_modules/@stdlib/number/float64/base/normalize/docs/types/index.d.ts:40]:
```suggestion
	*
	* var out = normalize( 3.14e-319 );
```

File: lib/node_modules/@stdlib/number/float64/base/normalize/docs/types/index.d.ts [hunk 2]
```diff
@@ -23,77 +23,86 @@
...
+	*
+	* @param x - input value
+	* @returns output array
+	*
+	* @example
+	* var pow = require( `@stdlib/math/base/special/pow` );
+	*
+	* var out = normalize( 3.14e-319 );
+	* // returns <Float64Array>[ 1.4141234400356668e-303, -52 ]
+	* var y = out[ 0 ];
```
Comment (kgryte) [inline lib/node_modules/@stdlib/number/float64/base/normalize/docs/types/index.d.ts:43]:
```suggestion
	*
	* var y = out[ 0 ];
```

File: lib/node_modules/@stdlib/number/float64/base/normalize/docs/types/index.d.ts [hunk 3]
```diff
@@ -23,77 +23,87 @@
...
+	* @example
+	* var out = new Float64Array( 2 );
+	*
+	* var v = normalize.assign( 3.14e-319, out, 1, 0 );
+	* // returns <Float64Array>[ 1.4141234400356668e-303, -52 ]
+	*
+	* var bool = ( v === out );
+	* // returns true
+	*/
+	assign( x: number, out: ArrayLike<number>, stride: number, offset: number ): ArrayLike<number>; // tslint-disable-line max-line-length
```
Comment (kgryte) [inline lib/node_modules/@stdlib/number/float64/base/normalize/docs/types/index.d.ts]:
```suggestion
	assign( x: number, out: ArrayLike<any>, stride: number, offset: number ): ArrayLike<number>; // tslint-disable-line max-line-length
```

The input array need not contain only numbers on input.

File: lib/node_modules/@stdlib/number/float64/base/normalize/docs/types/index.d.ts [hunk 4]
```diff
@@ -23,77 +23,87 @@
...
 * @param x - input value
 * @returns output array
 *
 * @example
 * var pow = require( `@stdlib/math/base/special/pow` );
-*
 * var out = normalize( 3.14e-319 );
-* // returns [ 1.4141234400356668e-303, -52 ]
-*
+* // returns <Float64Array>[ 1.4141234400356668e-303, -52 ]
```
Comment (kgryte) [inline lib/node_modules/@stdlib/number/float64/base/normalize/docs/types/index.d.ts:100]:
```suggestion
* // returns <Float64Array>[ 1.4141234400356668e-303, -52 ]
*
```

File: lib/node_modules/@stdlib/number/float64/base/normalize/docs/types/test.ts [hunk 5]
```diff
@@ -46,6 +44,67 @@ import normalize = require( './index' );
 	normalize( out, ( x: number ): number => x ); // $ExpectError
 }
 
+// The compiler throws an error if the function is provided an unsupported number of arguments...
+{
+	normalize(); // $ExpectError
+	normalize( 1.0, 2.0 ); // $ExpectError
+}
+
+// Attached to the main export is an `assign` method which returns an ArrayLike<number>...
```
Comment (kgryte) [inline lib/node_modules/@stdlib/number/float64/base/normalize/docs/types/test.ts]:
```suggestion
// Attached to the main export is an `assign` method which returns an array-like object containing numbers...
```

File: lib/node_modules/@stdlib/number/float64/base/normalize/docs/types/test.ts [hunk 6]
```diff
@@ -46,6 +44,67 @@ import normalize = require( './index' );
...
+	normalize.assign( true, out, 1, 0 ); // $ExpectError
+	normalize.assign( false, out, 1, 0 ); // $ExpectError
+	normalize.assign( '5', out, 1, 0 ); // $ExpectError
+	normalize.assign( null, out, 1, 0 ); // $ExpectError
+	normalize.assign( [], out, 1, 0 ); // $ExpectError
+	normalize.assign( {}, out, 1, 0 ); // $ExpectError
+	normalize.assign( ( x: number ): number => x, out, 1, 0 ); // $ExpectError
+}
+
+// The compiler throws an error if the `assign` method is provided a second argument which is not an ArrayLike<number>...
```
Comment (kgryte) [inline lib/node_modules/@stdlib/number/float64/base/normalize/docs/types/test.ts]:
```suggestion
// The compiler throws an error if the `assign` method is provided a second argument which is not an array-like object...
```

**Review by kgryte — APPROVED** [review]
LGTM.

@steff456 From my quick search, looks like we'll need to update

- `@stdlib/math/base/special/frexp`
- `@stdlib/math/base/special/ldexp`

Are there any other packages which are affected by the changes introduced in this PR?

**Comment by steff456** [comment]
I doubled checked and the only two packages that needed the update were `@stdlib/math/base/special/frexp` and `@stdlib/math/base/special/ldexp`.

File: lib/node_modules/@stdlib/math/base/special/frexp/lib/frexp.js [hunk 7]
```diff
@@ -90,7 +90,7 @@ function frexp( out, x ) {
 		return out;
 	}
 	// If `x` is subnormal, normalize it...
-	normalize( X, x );
+	normalize( x, X, 1, 0 );
```
Comment (kgryte) [inline lib/node_modules/@stdlib/math/base/special/frexp/lib/frexp.js:93]: Need to also update L25 from

```javascript
var normalize = require( '@stdlib/number/float64/base/normalize' );
```

to 

```javascript
var normalize = require( '@stdlib/number/float64/base/normalize' ).assign;
```

File: lib/node_modules/@stdlib/math/base/special/ldexp/lib/main.js [hunk 8]
```diff
@@ -106,7 +106,7 @@ function ldexp( frac, exp ) {
 		return frac;
 	}
 	// Normalize the input fraction:
-	normalize( FRAC, frac );
+	normalize( frac, FRAC, 1, 0 );
```
Comment (kgryte) [inline lib/node_modules/@stdlib/math/base/special/ldexp/lib/main.js:109]: Similar comment as for `frexp`.

File: lib/node_modules/@stdlib/number/float64/base/normalize/README.md [hunk 9]
```diff
@@ -91,6 +70,22 @@ out = normalize( NaN );
 // returns [ NaN, 0 ]
 ```
 
+#### normalize( x, out, stride, offset )
```
Comment (kgryte) [inline lib/node_modules/@stdlib/number/float64/base/normalize/README.md]:
```suggestion
#### normalize.assign( x, out, stride, offset )
```

File: lib/node_modules/@stdlib/number/float64/base/normalize/README.md [hunk 10]
```diff
@@ -91,6 +70,22 @@ out = normalize( NaN );
...
+#### normalize.assign( x, out, stride, offset )
+
+Returns a normal number `y` and exponent `exp` satisfying `x = y * 2^exp` and assigns results to a provided output array.
+
+```javascript
+var Float64Array = require( '@stdlib/array/float64' );
+
+var out = new Float64Array( 2 );
+
+var v = normalize( out, 3.14e-319, 1, 0);
```
Comment (kgryte) [inline lib/node_modules/@stdlib/number/float64/base/normalize/README.md]:
```suggestion
var v = normalize.assign( 3.14e-319, out, 1, 0);
```

**Review by kgryte — APPROVED** [review]
LGTM. Thanks, @steff456!

**Pull request merged**